gh-113317: Change how Argument Clinic lists converters

vstinner opened the pull request:
* Add a new create_python_parser_namespace() function for PythonParser to pass objects to executed code.
* In run_clinic(), list converters using 'converters' and 'return_converters' dictionarties.
* test_clinic: add 'object()' return converter.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-113317 -->
* Issue: gh-113317
<!-- /gh-issue-number -->

vstinner committed:
gh-113317: Change how Argument Clinic lists converters

* Add a new create_python_parser_namespace() function for
  PythonParser to pass objects to executed code.
* In run_clinic(), list converters using 'converters' and
  'return_converters' dictionarties.
* test_clinic: add 'object()' return converter.

vstinner committed:
Fix mypy: add type annotations

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,30 @@ def parse_file(
     libclinic.write_file(output, cooked)
 
 
+def create_python_parser_namespace() -> dict[str, Any]:
```
Why `python` in the name?

```suggestion
def create_parser_namespace() -> dict[str, Any]:
```

vstinner replied:
The function name is based on its lone user: the PythonParser class. I'm fine with dropping "python" in its name.

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,30 @@ def parse_file(
 class PythonParser:
     def __init__(self, clinic: Clinic) -> None:
-        pass
+        self.namespace = create_python_parser_namespace()
```
```suggestion
        self.namespace = create_parser_namespace()
```

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,30 @@ def parse_file(
+        self.namespace = create_python_parser_namespace()
 
     def parse(self, block: Block) -> None:
+        ns = dict(self.namespace)
```
It is already a dict, no?

vstinner replied:
It's to create a copy, since each exec() call can modify the newly created `ns` dict.

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -4984,25 +5001,22 @@ def run_clinic(parser: argparse.ArgumentParser, ns: argparse.Namespace) -> None:
-                    ids.append((name, name.removesuffix(suffix)))
-                    break
+        converter_list: list[tuple[str, str, Any]] = []
+        return_converter_list: list[tuple[str, str, Any]] = []
```
We have type aliases for converters and return converters; there should be no need to use `Any`.

erlend-aasland replied:
Alternatively, you could just use comprehensions?

vstinner replied:
I failed to please mypy: later, the two separated lists are consumed by the same loop which makes mypy really grumpy.

erlend-aasland replied:
I fixed it in 52da03c2a6d65b22c996527b03e0cccdd01b75ee

erlend-aasland replied:
Hm, I only fixed `ConverterType`. I failed to please mypy with `ReturnConverterType` :(

vstinner replied:
> I fixed it in https://github.com/python/cpython/commit/52da03c2a6d65b22c996527b03e0cccdd01b75ee

That was my first attempt, but mypy tests fail with this :-(

The correct fix is to use `ConverterType | ReturnConverterType`.

vstinner committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,30 @@ def parse_file(
+        ns[f'{name}_converter'] = converter
+    for name, return_converter in return_converters.items():
+        ns[f'{name}_return_converter'] = return_converter
+    return ns
```
I think we should make this a `PythonParser` method, instead of adding to the global namespace. What do you think?

vstinner replied:
I prefer to use a function, since I reused it in eval_ast_expr().

erlend-aasland committed:
Use proper types for converters and return converters

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -4984,25 +5001,22 @@ def run_clinic(parser: argparse.ArgumentParser, ns: argparse.Namespace) -> None:
+
+        for name, converter in converters.items():
+            converter_list.append((
+                f'{name}_converter',
```
This tuple item seems to be unused. Ditto for the return converter list.

vstinner committed:
Update

* Use also create_parser_namespace() in eval_ast_expr().
* Only createa the parser namespace once: add _BASE_PARSER_NAMESPACE.
* run_clinic(): remove the first tuple item in converter_list and
  return_converter_list. Rename the new first tuple item from
  'short_name' to 'name'.

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,35 @@ def parse_file(
     libclinic.write_file(output, cooked)
 
 
+def _create_parser_namespace() -> dict[str, Any]:
```
```suggestion
def _create_parser_namespace() -> dict[str, ConverterType | ReturnConverterType]:
```

vstinner replied:
Namespace keys have *many* types, I prefer to not have to list them :-( They are not only converters.

vstinner committed:
Fix create_parser_namespace()

* Add NoneType.
* Don't call create_parser_namespace() to startup, but only populate
  the cache at the first call.
* Remove eval_ast_expr() 'globals' argument.

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -4984,25 +5008,21 @@ def run_clinic(parser: argparse.ArgumentParser, ns: argparse.Namespace) -> None:
-                    break
+        any_converter_type = ConverterType | ReturnConverterType
+        converter_list: list[tuple[str, any_converter_type]] = []
+        return_converter_list: list[tuple[str, any_converter_type]] = []
```
```suggestion
        AnyConverter = ConverterType | ReturnConverterType
        converter_list: list[tuple[str, AnyConverter]] = []
        return_converter_list: list[tuple[str, AnyConverter]] = []
```

erlend-aasland replied:
... or `AnyConverterType`. At least follow the typing style convention :)

vstinner replied:
I changed the name for: AnyConverterType

vstinner committed:
AnyConverterType

erlend-aasland commented on Tools/clinic/clinic.py:
```diff
@@ -1988,13 +1988,38 @@ def parse_file(
     libclinic.write_file(output, cooked)
 
 
+def create_parser_namespace() -> dict[str, Any]:
```
Can you use `functools.cache` instead of adding custom cache code? `functools` is already imported.

vstinner replied:
Ok, done. I added `_create_parser_base_namespace()` function for that.

The function doesn't just return the same object at each call: each call creates a copy.

vstinner committed:
Use @functools.cache

erlend-aasland reviewed (approved):
Thanks for doing this; sorry it took so long to review!
